File: scripts/generate_silero_fixture.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.10"
# dependencies = ["silero-vad==5.1.2", "soundfile", "numpy", "onnxruntime", "torch"]
# ///
"""Golden per-window probabilities from the silero-vad package (ONNX mode).

Reference: silero_vad.load_silero_vad(onnx=True) called per 512-sample
window, exactly as the package's VADIterator does (including its 64-sample
context handling).
"""

import json
import os

import numpy as np
import soundfile as sf
import torch
from silero_vad import load_silero_vad

HERE = os.path.dirname(os.path.abspath(__file__))


def main():
    wav, sr = sf.read(
        os.path.join(HERE, "..", "tests", "assets", "hello_en.wav"), dtype="int16")
    assert sr == 16000
    pcm = (wav / 32768.0).astype(np.float32)

    model = load_silero_vad(onnx=True)
    model.reset_states()
    probs = []
    for i in range(0, len(pcm) - 511, 512):
        chunk = torch.from_numpy(pcm[i:i + 512])
        probs.append(float(model(chunk, 16000).item()))

    out = os.path.join(HERE, "..", "tests", "fixtures", "silero.json")
    with open(out, "w") as f:
        json.dump({"wav": "hello_en.wav", "windowSamples": 512, "probs": probs}, f)
    print(f"wrote {out} ({len(probs)} windows)")


if __name__ == "__main__":
    main()

File: scripts/generate_silero_fixture.ts
// Golden per-window probabilities for the Silero VAD provider.
//
// Reference: the silero_vad.onnx v5.1.2 model driven with onnxruntime-node
// exactly as the silero-vad package's OnnxWrapper does — 512-sample windows
// with the previous window's last 64 samples prepended as context, LSTM
// state threaded between calls. Wrapper semantics transcribed from:
// https://github.com/snakers4/silero-vad/blob/v5.1.2/src/silero_vad/utils_vad.py
// (class OnnxWrapper). Deliberately independent of src/providers/silero.ts
// so the fixture does not share the provider's implementation.
//
// Run: npm run fixtures

import { readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import * as ort from "onnxruntime-node";

import { readWav16kMono } from "../tests/wav.ts";

const WINDOW = 512;
const CONTEXT = 64;

const HERE = path.dirname(fileURLToPath(import.meta.url));

const wavBuf = readFileSync(path.join(HERE, "..", "tests", "assets", "hello_en.wav"));
const pcm = readWav16kMono(
  wavBuf.buffer.slice(wavBuf.byteOffset, wavBuf.byteOffset + wavBuf.byteLength),
);

const session = await ort.InferenceSession.create(
  path.join(HERE, "..", "models", "silero_vad.onnx"),
);
const sr = new ort.Tensor("int64", BigInt64Array.from([16000n]), []);
let state: ort.Tensor = new ort.Tensor("float32", new Float32Array(2 * 1 * 128), [2, 1, 128]);
let context = new Float32Array(CONTEXT);

const probs: number[] = [];
for (let i = 0; i + WINDOW <= pcm.length; i += WINDOW) {
  const window = pcm.subarray(i, i + WINDOW);
  const input = new Float32Array(CONTEXT + WINDOW);
  input.set(context);
  input.set(window, CONTEXT);
  const outputs = await session.run({
    input: new ort.Tensor("float32", input, [1, input.length]),
    state,
    sr,
  });
  const output = outputs.output;
  const stateN = outputs.stateN;
  if (output === undefined || stateN === undefined) {
    throw new Error("model did not return output and stateN");
  }
  state = stateN;
  context = window.slice(WINDOW - CONTEXT);
  probs.push((output.data as Float32Array)[0] ?? NaN);
}

const out = path.join(HERE, "..", "tests", "fixtures", "silero.json");
writeFileSync(out, JSON.stringify({ wav: "hello_en.wav", windowSamples: WINDOW, probs }) + "\n");
console.log(`wrote ${out} (${String(probs.length)} windows)`);
